src/resources/ble_esp32.cc

```diff
@@ -763,307 +1724,326 @@ PRIMITIVE(advertise_config) {
+  Object* error = object_to_mbuf(process, value, &om);
+  if (error) return error;
 
-  if (gatt->error() == BLE_HS_ENOENT) {
-    OUT_OF_RANGE;
-  } else if (gatt->error() != 0) {
+  uint32 flags = properties & 0x7F;
+  if (permissions & 0x1) flags |= BLE_GATT_CHR_F_READ;
+  if (permissions & 0x2 && !(properties & (BLE_GATT_CHR_PROP_WRITE | BLE_GATT_CHR_F_WRITE_NO_RSP)))
     INVALID_ARGUMENT;
-  }
-
+  if (permissions & 0x4) flags |= BLE_GATT_CHR_F_READ_ENC;
+  if (permissions & 0x8) flags |= BLE_GATT_CHR_F_WRITE_ENC;
 
-  return Primitive::integer(gatt->result(), process);
-}
+  BLECharacteristicResource* characteristic =
+    service_resource->get_or_create_characteristics_resource(
+        ble_uuid, flags,0,0,true);
 
-static Object* convert_mbuf_to_heap_object(Process* process, const os_mbuf* mbuf) {
-  int size = 0;
-  for (const os_mbuf* current = mbuf; current; current = SLIST_NEXT(current, om_next)) {
-    size += current->om_len;
-  }
-  ByteArray* data = process->object_heap()->allocate_internal_byte_array(size);
-  if (!data) return null;
-  ByteArray::Bytes bytes(data);
-  int offset = 0;
-  for (const os_mbuf* current = mbuf; current; current = SLIST_NEXT(current, om_next)) {
-    memmove(bytes.address() + offset, current->om_data, current->om_len);
-    offset += current->om_len;
+  if (!characteristic) {
+    if (om != null) os_mbuf_free(om);
+    MALLOC_FAILED;
   }
-  return data;
-}
 
+  if (om != null)
+    characteristic->set_mbuf_to_send(om);
 
-PRIMITIVE(request_data) {
-  ARGS(GATTResource, gatt);
+  proxy->set_external_address(characteristic);
+  return proxy;
+}
 
-  if (gatt->error() == BLE_HS_ENOENT) {
-    OUT_OF_RANGE;
-  } else if (gatt->error() != 0) {
-    INVALID_ARGUMENT;
-  }
+PRIMITIVE(add_descriptor) {
+  ARGS(BLECharacteristicResource, characteristic, Blob, raw_uuid, Object, value, int, properties, int, permissions)
 
-  const os_mbuf* mbuf = gatt->mbuf();
-  if (!mbuf) return process->program()->null_object();
-  Object* ret_val = convert_mbuf_to_heap_object(process, mbuf);
+  if (!characteristic->service()->peripheral_manager()) INVALID_ARGUMENT;
 
-  if (ret_val != null) {
-    gatt->set_mbuf(null);
-    return ret_val;
-  } else {
-    ALLOCATION_FAILED;
-  }
-}
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
 
-PRIMITIVE(send_data) {
-  ARGS(GATTResource, gatt, uint16, handle, Object, value);
+  ble_uuid_any_t ble_uuid = uuid_from_blob(raw_uuid);
 
   os_mbuf* om = null;
   Object* error = object_to_mbuf(process, value, &om);
   if (error) return error;
 
-  int err = ble_gattc_write(gatt->handle(), handle, om, NULL, NULL);
-  if (err != ESP_OK) {
-    return Primitive::os_error(err, process);
-  }
-
-  return process->program()->null_object();
-}
-
-PRIMITIVE(request_service) {
-  ARGS(GATTResource, gatt, Blob, uuid);
+  uint8 flags = 0;
+  if (properties & BLE_GATT_CHR_F_READ || permissions & 0x01) flags |= BLE_ATT_F_READ;
```

Comment:
Confusing that the order in the condition is inverted (compared to add_characteristic). In the other one we first look at `permissions`.